// ejv/build/cjs/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ErrorMsg = exports.ErrorType = exports.StringFormat = exports.NumberFormat = exports.DataType = void 0;
var DataType;
(function (DataType) {
    DataType["BOOLEAN"] = "boolean";
    DataType["NUMBER"] = "number";
    DataType["STRING"] = "string";
    DataType["OBJECT"] = "object";
    DataType["DATE"] = "date";
    DataType["REGEXP"] = "regexp";
    DataType["ARRAY"] = "array";
})(DataType || (exports.DataType = DataType = {}));
var NumberFormat;
(function (NumberFormat) {
    NumberFormat["INTEGER"] = "integer";
    NumberFormat["INDEX"] = "index";
})(NumberFormat || (exports.NumberFormat = NumberFormat = {}));
var StringFormat;
(function (StringFormat) {
    StringFormat["EMAIL"] = "email";
    StringFormat["DATE"] = "date";
    StringFormat["DATE_TIME"] = "date-time";
    StringFormat["TIME"] = "time";
    // URL = 'url',
    // IP = 'ip', // ipv4 || ipv6
    // IPV4 = 'ipv4',
    // IPV6 = 'ipv6'
})(StringFormat || (exports.StringFormat = StringFormat = {}));
var ErrorType;
(function (ErrorType) {
    // about ejv() running
    ErrorType["NO_DATA"] = "NO_DATA";
    ErrorType["NO_SCHEME"] = "NO_SCHEME";
    ErrorType["INVALID_SCHEMES"] = "INVALID_SCHEMES";
    // not match with scheme
    ErrorType["REQUIRED"] = "REQUIRED";
    ErrorType["TYPE_MISMATCH"] = "TYPE_MISMATCH";
    ErrorType["TYPE_MISMATCH_ONE_OF"] = "TYPE_MISMATCH_ONE_OF";
    ErrorType["ONE_VALUE_OF"] = "ONE_VALUE_OF";
    ErrorType["FORMAT"] = "FORMAT";
    ErrorType["FORMAT_ONE_OF"] = "FORMAT_ONE_OF";
    ErrorType["BIGGER_THAN_OR_EQUAL"] = "BIGGER_THAN_OR_EQUAL";
    ErrorType["BIGGER_THAN"] = "BIGGER_THAN";
    ErrorType["SMALLER_THAN_OR_EQUAL"] = "SMALLER_THAN_OR_EQUAL";
    ErrorType["SMALLER_THAN"] = "SMALLER_THAN";
    ErrorType["LENGTH"] = "LENGTH";
    ErrorType["MIN_LENGTH"] = "MIN_LENGTH";
    ErrorType["MAX_LENGTH"] = "MAX_LENGTH";
    ErrorType["PATTERN"] = "PATTERN";
    ErrorType["PATTERN_ONE_OF"] = "PATTERN_ONE_OF";
    ErrorType["PROPERTY"] = "PROPERTY";
    ErrorType["AFTER_OR_SAME_DATE"] = "AFTER_OR_SAME_DATE";
    ErrorType["AFTER_DATE"] = "AFTER_DATE";
    ErrorType["BEFORE_OR_SAME_DATE"] = "BEFORE_OR_SAME_DATE";
    ErrorType["BEFORE_DATE"] = "BEFORE_DATE";
    ErrorType["UNIQUE_ITEMS"] = "UNIQUE_ITEMS";
    ErrorType["ITEMS_TYPE"] = "ITEMS_TYPE";
    ErrorType["ITEMS_SCHEMES"] = "ITEMS_SCHEMES";
})(ErrorType || (exports.ErrorType = ErrorType = {}));
var ErrorMsg;
(function (ErrorMsg) {
    // about Error
    ErrorMsg["NO_DATA"] = "no data to validate";
    ErrorMsg["NO_SCHEME"] = "no scheme";
    ErrorMsg["NO_ARRAY_SCHEME"] = "schemes should be array";
    ErrorMsg["NO_OBJECT_ARRAY_SCHEME"] = "schemes should be array of object";
    ErrorMsg["EMPTY_SCHEME"] = "schemes should not be empty";
    ErrorMsg["SCHEMES_SHOULD_HAVE_TYPE"] = "schemes should have type";
    ErrorMsg["SCHEMES_HAS_INVALID_TYPE"] = "schemes has invalid type: <<1>>";
    ErrorMsg["SCHEMES_HAS_DUPLICATED_TYPE"] = "schemes has duplicated type: <<1>>";
    ErrorMsg["ENUM_SHOULD_BE_ARRAY"] = "enum should be array";
    ErrorMsg["ENUM_SHOULD_BE_NUMBERS"] = "enum values should be numbers";
    ErrorMsg["ENUM_SHOULD_BE_STRINGS"] = "enum values should be strings";
    ErrorMsg["MIN_SHOULD_BE_NUMBER"] = "min should be number";
    ErrorMsg["EXCLUSIVE_MIN_SHOULD_BE_BOOLEAN"] = "exclusiveMin should be a boolean";
    ErrorMsg["MAX_SHOULD_BE_NUMBER"] = "max should be number";
    ErrorMsg["EXCLUSIVE_MAX_SHOULD_BE_BOOLEAN"] = "exclusiveMax should be a boolean";
    ErrorMsg["INVALID_NUMBER_FORMAT"] = "invalid number format: <<1>>";
    ErrorMsg["INVALID_STRING_FORMAT"] = "invalid string format: <<1>>";
    ErrorMsg["INVALID_STRING_PATTERN"] = "invalid string pattern: <<1>>";
    ErrorMsg["LENGTH_SHOULD_BE_INTEGER"] = "length should be a integer";
    ErrorMsg["MIN_LENGTH_SHOULD_BE_INTEGER"] = "minLength should be a integer";
    ErrorMsg["MAX_LENGTH_SHOULD_BE_INTEGER"] = "maxLength should be a integer";
    ErrorMsg["PROPERTIES_SHOULD_BE_ARRAY"] = "properties should be array";
    ErrorMsg["PROPERTIES_SHOULD_HAVE_ITEMS"] = "properties should have items";
    ErrorMsg["PROPERTIES_SHOULD_BE_ARRAY_OF_OBJECT"] = "properties should be array of object";
    ErrorMsg["ALLOW_NO_PROPERTY_SHOULD_BE_BOOLEAN"] = "allowNoProperty should be a boolean";
    ErrorMsg["MIN_DATE_SHOULD_BE_DATE_OR_STRING"] = "min should be Date or string representing date";
    ErrorMsg["MAX_DATE_SHOULD_BE_DATE_OR_STRING"] = "max should be Date or string representing date";
    ErrorMsg["UNIQUE_SHOULD_BE_BOOLEAN"] = "unique should be a boolean";
    ErrorMsg["INVALID_ITEMS_SCHEME"] = "invalid schemes of array items: <<1>>";
    // about EjvError
    ErrorMsg["REQUIRED"] = "value is required";
    ErrorMsg["TYPE_MISMATCH"] = "the value should be a <<1>>";
    ErrorMsg["TYPE_MISMATCH_ONE_OF"] = "the value should be type of <<1>>";
    ErrorMsg["ONE_VALUE_OF"] = "the value should be one of <<1>>";
    ErrorMsg["FORMAT"] = "the value should be a form of <<1>>";
    ErrorMsg["FORMAT_ONE_OF"] = "the value should be form of one of <<1>>";
    ErrorMsg["BIGGER_THAN_OR_EQUAL"] = "the value should be bigger or equal than <<1>>";
    ErrorMsg["BIGGER_THAN"] = "the value should be bigger than <<1>>";
    ErrorMsg["SMALLER_THAN_OR_EQUAL"] = "the value should be smaller or equal than <<1>>";
    ErrorMsg["SMALLER_THAN"] = "the value should be smaller than <<1>>";
    ErrorMsg["LENGTH"] = "the value should be length of <<1>>";
    ErrorMsg["MIN_LENGTH"] = "the value should be longer than <<1>>";
    ErrorMsg["MAX_LENGTH"] = "the value should be shorter than <<1>>";
    ErrorMsg["PATTERN"] = "the value should be pattern of <<1>>";
    ErrorMsg["PATTERN_ONE_OF"] = "the value should be one of pattern of <<1>>";
    ErrorMsg["PROPERTY"] = "the value should have property";
    ErrorMsg["AFTER_OR_SAME_DATE"] = "the value should be after or same date than <<1>>";
    ErrorMsg["AFTER_DATE"] = "the value should be after date than <<1>>";
    ErrorMsg["BEFORE_OR_SAME_DATE"] = "the value should be before or same date than <<1>>";
    ErrorMsg["BEFORE_DATE"] = "the value should be before date than <<1>>";
    ErrorMsg["UNIQUE_ITEMS"] = "the array should be unique items";
    ErrorMsg["ITEMS_TYPE"] = "the array should have items in type of <<1>>";
    ErrorMsg["ITEMS_SCHEMES"] = "the array should have items matched with schemes of <<1>>";
})(ErrorMsg || (exports.ErrorMsg = ErrorMsg = {}));
