// ejv/build/cjs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EjvError = exports.ErrorType = exports.ErrorMsg = exports.StringFormat = exports.NumberFormat = exports.DataType = exports.ejv = void 0;
var ejv_1 = require("./ejv.js");
Object.defineProperty(exports, "ejv", { enumerable: true, get: function () { return ejv_1.ejv; } });
var constants_1 = require("./constants.js");
Object.defineProperty(exports, "DataType", { enumerable: true, get: function () { return constants_1.DataType; } });
Object.defineProperty(exports, "NumberFormat", { enumerable: true, get: function () { return constants_1.NumberFormat; } });
Object.defineProperty(exports, "StringFormat", { enumerable: true, get: function () { return constants_1.StringFormat; } });
Object.defineProperty(exports, "ErrorMsg", { enumerable: true, get: function () { return constants_1.ErrorMsg; } });
Object.defineProperty(exports, "ErrorType", { enumerable: true, get: function () { return constants_1.ErrorType; } });
var interfaces_1 = require("./interfaces.js");
Object.defineProperty(exports, "EjvError", { enumerable: true, get: function () { return interfaces_1.EjvError; } });
